feat(codegen): support for api key auth trait

eduardomourar opened the pull request:
## Motivation and Context
This adds codegen support for services and operations that use the `@httpApiKeyAuth trait`. It depends on: https://github.com/awslabs/smithy-rs/pull/2153.

## Description
It will perform the proper mutation of the HTTP request to include the API Key. The generated client will have that logic automatically included. This value will be injected into the header/query parameter that's specified in the trait.

## Testing
Tested only locally on our own Smithy models, but I will add some unit tests before merging it into main.

## Checklist
<!--- If a checkbox below is not applicable, then please DELETE it rather than leaving it unchecked -->
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,220 @@
+ * @param codegenContext Codegen context that includes the model and service shape
+ * @return if the httpApiKeyAuth trait is used by the service
+ */
+fun isSupportedApiKeyAuth(codegenContext: ClientCodegenContext): Boolean {
```
```suggestion
private fun isSupportedApiKeyAuth(codegenContext: ClientCodegenContext): Boolean {
```

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,220 @@
+ * @param operation operation shape
+ * @return if the service and operation have the httpApiKeyAuthTrait
+ */
+fun hasApiKeyAuthScheme(codegenContext: ClientCodegenContext, operation: OperationShape): Boolean {
```
```suggestion
private fun hasApiKeyAuthScheme(codegenContext: ClientCodegenContext, operation: OperationShape): Boolean {
```

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,220 @@
+        }
+}
+
+fun apiKey(runtimeConfig: RuntimeConfig) = RuntimeType.smithyHttpAuth(runtimeConfig).resolve("api_key::AuthApiKey")
```
```suggestion
private fun apiKey(runtimeConfig: RuntimeConfig) = RuntimeType.smithyHttpAuth(runtimeConfig).resolve("api_key::AuthApiKey")
```

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,220 @@
+    override fun section(section: LibRsSection): Writable = when (section) {
+        is LibRsSection.Body -> writable {
+            rust(
+                "pub use #T;",
```
`Credentials` is currently in the crate root, but that will change with [RFC-26](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0026_client_crate_organization.md) (which I'm currently implementing). Could you make this an `extras` customization and place this re-export in the `crate::config` module instead?

Something like:
```kotlin
    override fun extras(codegenContext: ClientCodegenContext, rustCrate: RustCrate) {
        rustCrate.withModule(ClientRustModule.Config) {
            rust("pub use #T;", apiKey(runtimeConfig))
        }
    }
```
On `ApiKeyAuthDecorator`

eduardomourar replied:
done

jdisanti commented on codegen-client/src/test/kotlin/software/amazon/smithy/rust/codegen/client/customizations/ApiKeyAuthDecoratorTest.kt:
```diff
@@ -0,0 +1,81 @@
+    """.asSmithyModel()
+
+    @Test
+    fun `set an api key in the property bag`() {
```
I'd like to also see separate tests that exercise the auth header getting set in the query string and on the authorization header.

I think these should be set on the returned `operation` from the `make_operation` call.

jdisanti commented on codegen-client-test/model/rest-xml-extras.smithy:
```diff
@@ -10,6 +10,8 @@ use smithy.test#httpRequestTests
 /// A REST XML service that sends XML requests and responses.
 @service(sdkId: "Rest XML Protocol")
 @restXml
+@httpApiKeyAuth(scheme: "Apikey", name: "Authorization", in: "header")
+@auth([httpApiKeyAuth])
```
I think the Kotlin unit test should be sufficient to test this, so I don't think this is necessary.

eduardomourar replied:
removed

jdisanti reviewed (commented):
This looks mostly good. I kicked off the codegen diff bot to see what the generated code diff looks like (hopefully it works this time): https://github.com/awslabs/smithy-rs/actions/runs/4179922718

Some small comments about organization and testing.

Velfi commented:
@eduardomourar ping me or @jdisanti when you're ready for another review.

eduardomourar commented:
> @eduardomourar ping me or @jdisanti when you're ready for another review.

@Velfi, I have addressed all comments. Let me know if there is anything additional to be done.

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,204 @@
+            is ServiceConfig.BuilderImpl -> writable {
+                rustTemplate(
+                    """
+                    /// Sets the api key that will be used by the client.
```
```suggestion
                    /// Sets the API key that will be used by the client.
```

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,204 @@
+                        self
+                    }
+
+                    /// Sets the api key that will be used by the client.
```
```suggestion
                    /// Sets the API key that will be used by the client.
```

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,204 @@
+            is ServiceConfig.ConfigImpl -> writable {
+                rustTemplate(
+                    """
+                    /// Returns api key used by the client, if it was provided.
```
```suggestion
                    /// Returns API key used by the client, if it was provided.
```

jdisanti commented on codegen-client/src/test/kotlin/software/amazon/smithy/rust/codegen/client/customizations/ApiKeyAuthDecoratorTest.kt:
```diff
@@ -0,0 +1,175 @@
+                            .await
+                            .expect("valid operation");
+                        assert_eq!(
+                            op.request().uri().query(),
```
This might fix the test compile:
```suggestion
                            operation.request().uri().query(),
```

jdisanti commented on codegen-client/src/test/kotlin/software/amazon/smithy/rust/codegen/client/customizations/ApiKeyAuthDecoratorTest.kt:
```diff
@@ -0,0 +1,175 @@
+                            &AuthApiKey::new(api_key_value),
+                        );
+                        assert_eq!(
+                            op.request().headers().contains_key("authorization"),
```
```suggestion
                            operation.request().headers().contains_key("authorization"),
```

jdisanti reviewed (commented):
This is looking good! I'm just about ready to approve and merge. The new tests don't appear to be passing due to a compile error, but they look exactly like what I wanted. I added some code suggestions that might fix them.

jdisanti commented on codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/ApiKeyAuthDecorator.kt:
```diff
@@ -0,0 +1,204 @@
+        return baseCustomizations
+    }
+
+    override fun extras(codegenContext: ClientCodegenContext, rustCrate: RustCrate) {
```
It looks like this is running without checking `applies`, which is resulting in the re-export getting added to the SDK, and that is what's causing CI to fail currently.
